tests/subsys/net/lib/mcumgr_smp_client/src/main.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+#include <mcumgr_smp_client.h>
+#include <dfu/dfu_target_smp.h>
+#include "fota_download_util.h"
+
+static struct fota_download_evt fota_cb_event;
+static fota_download_callback_t client_test_callback;
+static bool scheduled_called;
+static bool reboot_called;
+static bool dfu_target_done_success;
+static bool dfu_target_image_confirmed;
+struct mcumgr_image_data image_list[2];
+
+/* Stubbed Functions start */
+int dfu_target_smp_recovery_mode_enable(dfu_target_reset_cb_t cb)
+{
+	return 0;
+}
+
+int dfu_target_smp_client_init(void)
+{
+	return 0;
+}
+
+bool nrfx_is_in_ram(char *const file)
+{
+	return true;
+}
+
+int dfu_target_init(int img_type, int img_num, size_t file_size, dfu_target_callback_t cb)
+{
+	return 0;
+}
+
+void fota_download_test_callback(const struct fota_download_evt *evt)
+{
+	fota_cb_event.id = evt->id;
+	if (evt->id == FOTA_DOWNLOAD_EVT_ERROR) {
+		fota_cb_event.cause = evt->cause;
+	} else if (evt->id == FOTA_DOWNLOAD_EVT_PROGRESS) {
+		fota_cb_event.progress = evt->progress;
+	}
+
+	printf("Fota CB %d\r\n", fota_cb_event.id);
+}
+
+int fota_download_start_with_image_type(const char *host, const char *file, int sec_tag,
+					uint8_t pdn_id, size_t fragment_size,
+					const enum dfu_target_image_type expected_type)
+{
+
+	return 0;
+}
+
+int fota_download_util_client_init(fota_download_callback_t client_callback, bool smp_image_type)
+{
+	client_test_callback = client_callback;
+	return 0;
+}
+
+int fota_download_cancel(void)
+{
+	struct fota_download_evt fota_event;
+
+	if (client_test_callback) {
+		fota_event.id = FOTA_DOWNLOAD_EVT_CANCELLED;
+		client_test_callback(&fota_event);
+	}
+
+	return 0;
+}
+
+int dfu_target_schedule_update(int img_num)
+{
+	scheduled_called = true;
+	image_list[1].flags.pending = true;
+	return 0;
+}
+
+int dfu_target_smp_reboot(void)
+{
+	reboot_called = true;
+	return 0;
+}
+
+int dfu_target_smp_done(bool succesufully)
+{
+	dfu_target_done_success = succesufully;
+	return 0;
+}
+
+int dfu_target_reset(void)
+{
+	return dfu_target_smp_done(false);
+}
+
+int dfu_target_smp_confirm_image(void)
+{
+	dfu_target_image_confirmed = true;
+	image_list[0].flags.confirmed = true;
+	return 0;
+}
+
+int dfu_target_reset_smp_test_cb(void)
+{
+	return 0;
+}
+
+int dfu_target_smp_image_list_get(struct mcumgr_image_state *res_buf)
+{
+	res_buf->status = MGMT_ERR_EOK;
+	res_buf->image_list_length = 2;
+	res_buf->image_list = image_list;
+	return 0;
+}
+
+/* Stubbed Functions end */
+
+ZTEST(mcumgr_smp_client_test, test_init)
```

Comment:
`Ztest` framework executes tests in alphanumerical order. At first glance it seems to me that this `init` test should be executed at the beginning (but I didn't analyze it deeper, so maybe I'm wrong). When I run those tests locally I can see that test execution order is as below:
```
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_download] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_erase_and_confirm] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_init] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_reset] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_schedule_image] duration = 0.000 seconds
```
`test_init` test is executed as a third test. Is it intentionally? If not, maybe you should consider changing this `test_init` test as a `ztest_suite_setup_t` function to guarantee that those actions will be executed always at the beginning of test suite.

Reply:
I can add init to ZTEST_SUITE() pre action function so Then it will be done properly. Good point. I can change that easily.